src/main/java/io/cryostat/core/reports/InterruptibleReportGenerator.java

```diff
@@ -0,0 +1,350 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.RunnableFuture;
+import java.util.stream.Collectors;
+
+import org.openjdk.jmc.common.io.IOToolkit;
+import org.openjdk.jmc.common.item.IItemCollection;
+import org.openjdk.jmc.common.util.IPreferenceValueProvider;
+import org.openjdk.jmc.common.util.XmlToolkit;
+import org.openjdk.jmc.flightrecorder.CouldNotLoadRecordingException;
+import org.openjdk.jmc.flightrecorder.JfrLoaderToolkit;
+import org.openjdk.jmc.flightrecorder.rules.IRule;
+import org.openjdk.jmc.flightrecorder.rules.Result;
+import org.openjdk.jmc.flightrecorder.rules.RuleRegistry;
+import org.openjdk.jmc.flightrecorder.rules.report.html.internal.HtmlResultGroup;
+import org.openjdk.jmc.flightrecorder.rules.report.html.internal.HtmlResultProvider;
+import org.openjdk.jmc.flightrecorder.rules.report.html.internal.RulesHtmlToolkit;
+
+import io.cryostat.core.log.Logger;
+
+import org.jsoup.Jsoup;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+/**
+ * Re-implementation of {@link ReportGenerator} where the report generation task is represented by a
+ * {@link Future}, allowing callers to cancel ongoing rules report analyses or to easily time out on
+ * analysis requests. This should eventually replace {@link ReportGenerator} entirely - there should
+ * only be benefits to using this implementation.
+ */
+public class InterruptibleReportGenerator {
+
+    private final Logger logger;
+    private final Set<ReportTransformer> transformers;
+
+    private final ExecutorService executor;
+    private final ExecutorService qThread = Executors.newCachedThreadPool();
+
+    public InterruptibleReportGenerator(
+            Logger logger, Set<ReportTransformer> transformers, ExecutorService executor) {
+        this.logger = logger;
+        this.transformers = transformers;
+        this.executor = executor;
+    }
+
+    public Future<String> generateReportInterruptibly(InputStream recording) {
+        return qThread.submit(
+                () -> {
+                    // this is generally a re-implementation of JMC JfrHtmlRulesReport#createReport,
+                    // but calling our cancellable evalute() method rather than the
+                    // RulesToolkit.evaluateParallel as explained further down.
+                    List<Future<Result>> resultFutures = new ArrayList<>();
+                    try {
+                        resultFutures.addAll(
+                                evaluate(
```

Comment:
I realize this has been already merged but I have a couple of questions.

So when `evaluate` is called with the list of `rules` in the registry and the actual `events` content of the recording in question, the recording is evaluated against each `rule` and packaged into a `RunnableFuture<Result>`, which gets `run` and the `Result` from calling `get` packaged into a `Callable<Result>`. A list of these is returned, streamed, and mapped to `executor::submit`, which spits out a cancellable `Future<Result>`.

What I don't understand is didn't we already wait for the rule evaluation to complete inside `evaluate` itself, specifically when `return result.get()` was called on line 236?

Also, am I correct in saying that the way report generation works is by evaluating the recording against a bunch of standard rules for proper JVM behaviour, the results for which (i.e. any errors) are then mapped to certain result groups, which allows for a clear, organized report once everything has been processed to HTML, correct?

Reply:
Actually, I thought about it more and realized that when the lambda function inside `evaluate` is packaged into a `Callable<Result>`, it happens asynchronously so I think the actual rule evaluation isn't triggered until `map(executor::submit)` is called on line 113.

Reply:
On line 236 we are doing a blocking `.get()` call from within a task submitted to the thread pool, so these are being processed asynchronously on a different thread than the main one that is first handling the request. If there is only one CPU core available then this basically just degenerates to doing it sequentially, but it's still a separate thread, and so we can interrupt it and have the CPU context-switch back to the main thread and stop processing any rules.

> Also, am I correct in saying that the way report generation works is by evaluating the recording against a bunch of standard rules for proper JVM behaviour, the results for which (i.e. any errors) are then mapped to certain result groups, which allows for a clear, organized report once everything has been processed to HTML, correct?

Yep, pretty much. If you look at an HTML rules report, each of those coloured score bar categories that you see is a rule. If you have enough CPU cores then all of those can be evaluated in parallel. Once they are evaluated then the result score is used to generate the HTML for the coloured bar, the associated text is placed next to it, related rules are grouped together, and all of those are finally combined into the full page HTML report document.

Reply:
Or rather - on line 236 we are converting the `RunnableFuture` representing the `Rule` into a simpler `Callable`, which both performs the execution of the rule and returns its result. Line 113, `executor::submit`, is what actually turns those `Callables` into asynchronous tasks processed by the thread pool.

Reply:
Awesome, makes sense. Thanks!